### ISSUE TITLE
CSV file of SDM model outputs

### ISSUE THREAD

**Issue by johnvanbreda** [BiologicalRecordsCentre/irecord-extras#2]
CEH to provide a CSV file or similar for the output of SDM model runs:
* preferred name TVK
* grid square
* lat and long coordinate of grid square centre
* week number - this can be blank for taxa where the probability does not change throughout the year
* probability (0 to 1 or 0 to 100?) 
* certainty
* model run ID

**Comment by johnvanbreda** [comment]
@DavidRoy Just a thought - if it is easier to directly connect Logstash to a database query or some other output, then it should be possible to configure Logstash to do that rather than use a CSV file.
Also please let me know what format you expect the probability and certainty fields to take.

**Comment by DavidRoy** [comment]
@robboyd as discussed, can you supply an example .csv?

@johnvanbreda i think .csv is probably easiest as we won't update these models very frequently.

tagging @sacrevert for info

**Comment by johnvanbreda** [comment]
A comment from an email thread which needs to be resolved before finalising the model output:
"I think the week number should be null if a taxon is equally probable throughout the year as that’s more logical. Or we could create 52 entries but that seems overkill. Another thought would be instead of using a week number, we add fields for start day and end day, with values from 1-365. Then week based entries would be given a start and end 7 days apart. Or you could have an entry which covers a larger range of days (the whole summer for example), or in this case, the whole year. This approach would be more accurate as not prone to variability depending on the day the 1st of January falls on and would also allow more granular data when necessary, or to reduce the number of entries required for less variable data."

**Comment by robboyd** [comment]
SDM outputs are currently in the form of .asc files on the British grid (defined in eastings and northings). Will take some processing to get the .csv you're after. It will be relatively straightforward to give you lat/lon of centre of the cell, probability of occurrence (note this static in time, so no need for e.g. week field) and grid reference. Currently working on mapping the uncertainty, and don't know how to get TVKs. Some species names I have are in BRC concept code format and I'm working on translating them to latin for something else. How quicjkly do you need an example?

**Comment by sacrevert** [comment]
Yes, preferred TVK definitely the most troublesome element here. I think that myself, with support from Colin, will have to gradually sort this out, possible as we are going through the validation process.

**Comment by DavidRoy** [comment]
@sacrevert shall we start with the vascular plants (and bryophytes) as the TVK link might be easier?
@robboyd the uncertainty is not so important at this stage
@johnvanbreda in most cases we won't have a seasonality.  Probably only butterflies and moths to start with.  I do wonder if we should have a life-stage as part of the data model.  Adult in most cases.  Am thinking of butterflies where most data comes from egg counts (e.g. black hairstreak)

**Comment by sacrevert** [comment]
Yes, the bryos are straightforward, and vasc plants reasonably so. Not sure what timescale this is being done to, but presumably someone will let me know when they need them.

**Comment by johnvanbreda** [comment]
@DavidRoy understood, though the "What's flying tonight" output will be very dependent on seasonality, so it's important to plan for it from the outset. If we go with the start day + end day approach, then it can be set to 1 - 366 for all non-seasonal data.

@robboyd the specification of a CSV file is just a suggestion - it may be possible to extract data from other file formats. What is an ASC file - can you supply a small example? 

As to species names, it needs to be something we can easily map to the equivalent species concept in UKSI. The accepted scientific name should work OK in the vast majority of cases. If you like, you could export as BRC concept codes then just provide me with a lookup CSV file. Does BRC have a lookup from BRC code to taxon version key, or organism key, or some other UKSI key?

**Comment by johnvanbreda** [comment]
@sacrevert this is one of several projects I'm working on at the moment which I'm aiming to do by end March, if that gives you some idea of timescales.

**Comment by sacrevert** [comment]
@johnvanbreda bryophytes and vascular plants are easy, but I have no idea about whether the end of March is a realistic timescale for the invert groups. The SDM data were collated in different ways,and have different identifiers, so it's not just a case of a lookup (if we had a lookup it would be trivial for us to provide TVK). Probably we'll have to do something approximate in some groups and live with the inevitable errors until a recorder or whoever points them out.

**Comment by johnvanbreda** [comment]
Ok, thanks @sacrevert - I think if we can at least establish the principle of the SDM data store using bryophytes and vascular plants it will be a good start - we can add other groups later. If you are able to provide some examples of the taxonomic concept identifiers you do hold in the SDM outputs, such as species names or codes, or whatever, then I'm happy to have a look at whether there is an easy way to map them to UKSI.

**Comment by DavidRoy** [comment]
@johnvanbreda I've sent you an example SDM file

**Comment by DavidRoy** [comment]
@johnvanbreda we are ready to send more files so we can get to the point of testing the API.  Is the file format ok?

**Comment by sacrevert** [comment]
Minor point, but can we name the 'probability' field 'score', rather than probability, as it's not a true probability, and calling it so may lead to confusion and/or misuse. @johnvanbreda @robboyd @DavidRoy

**Comment by johnvanbreda** [comment]
@DavidRoy in the example file I have the following:
* Grid ref
* Suitability
* Species
* Group

In the original suggestion at the top of this thread we had the following:
* preferred name TVK
* grid square
* lat and long coordinate of grid square centre
* week number - this can be blank for taxa where the probability does not change throughout the year
* probability (0 to 1 or 0 to 100?)
* certainty
* model run ID

As discussed elsewhere, the TVK can be calculated during import. Shall we rename suitability to "Score" and use that instead of probability?

Do we need to add model run ID?

Is it possible to add lat and long coordinate of grid square centre?

One last thing - we will need an Elasticsearch instance to host this on. We could potentially host it on the existing live ES instance but I would be concerned about putting too much into the 1 machine. Can we have another VM to run an SDM Elasticsearch on please?

**Comment by sacrevert** [comment]
Happy to let @robboyd decide between suitability and score. relativeSuitability might be even better.

Rob should be able to add lat/long of square centroid using functions in BRCmap.

**Comment by robboyd** [comment]
relativeSuitability is best for sure. Easy to add coords of the centroids as @sacrevert says. When do you need this by?

**Comment by DavidRoy** [comment]
@robboyd can you export the ant results in this format sometime next week

@BirenRathod can you ask CCS to setup a new VM so we can install an SDM ElasticSearch

**Comment by robboyd** [comment]
No problem @DavidRoy
